### PR TITLE
Run `crate-ci/typos` in CI

### PR DESCRIPTION
**Pull request by LilithHafner**
https://github.com/crate-ci/typos

TODO:
- [x] In the `actions/checkout` action, set `persist-credentials: false`.
- [x] Set a short timeout (e.g. 2 minutes or 5 minutes).
- [x] Fix true spelling errors (#51709)
~- [ ] Create an ignore list with the false positives?~

<details>
<summary>List of false positives that could be stored in typos.toml, if folks decide to go that route down the line</summary>

```toml
# Hello and Welcome to typos.toml!

## Policy
# tldr: Don't let CI bully you. Use whatever varaible names you want.
#
# The spell check CI feature exists to help detect accidental typos, not to constrain the
# language we use in this repo. If you want to use a word or a variable name or somesuch
# that the CI check is not happy with, go right ahead, just make sure to add the appropriate
# entry to this file to tell the CI check that you didn't make a mistake. As a reviewier, if
# you think that a variable name is bad, go right ahead and tell the author that, just don't
# cite the CI check as a reason, that's not what the check is for.

## How to
# tldr: Add the whole identifier right below [default.extend-identifiers]
#
# There are two magor sections in this file, [default.extend-identifiers] (near the top) and
# [default.extend-words] (near the bottom). If the false positive is a full identifier like
# `mrs_claus` or `breal`, add it to [default.extend-identifiers] and that specific
# identifier will no longer be flagged as a typo. If, on the other hand, it is a word that
# appeasrs inside multiple a multi-word identifiers, like `ser` which is used in
# `jl_lookup_ser_tag`, `ser_tag`, `ser_version`, etc. then insrted of adding each of those
# identifiers individually, simply add it to the [default.extend-words] section. Identifiers
# are case sensitive and words are not.

[default.extend-identifiers]
mrs_claus = "mrs_claus"
womens = "womens"
breal = "breal"
Breal = "Breal"
Numer = "Numer"
splitted = "splitted"
affinitized = "affinitized"
key_smove = "key_smove"
readed_zero = "readed_zero"
check_inconsistentcy = "check_inconsistentcy"
L_OP_CALLL = "L_OP_CALLL"
OP_CALLL = "OP_CALLL"

# strings in test directories
Facilisi = "Facilisi"
vailable = "vailable"

# typos that are programmatically observable
supress_output = "supress_output"
hashs_seed = "hashs_seed"

[default.extend-words]
ba = "ba"
egal = "egal"
parm = "parm"
modul = "modul"
nd = "nd"
strat = "strat"
WRONLY = "WRONLY"
thisy = "thisy"
vally = "vally"
Ot = "Ot"
clos = "clos"
applys = "applys"
findn = "findn"
mis = "mis"
seh = "seh"
ue = "ue"
Missings = "Missings"
SOM = "SOM"
sais = "sais"
Merly = "Merly"
Filetimes = "Filetimes"
HSA = "HSA"
Strategems = "Strategems"
anumber = "anumber"
uupper = "uupper"
shttp = "shttp"
ser = "ser"
somes = "somes"
sav = "sav"
egals = "egals"
OLT = "OLT"
eyt = "eyt"
matc = "matc"
nam = "nam"
alls = "alls"
noe = "noe"

# strings in tests directories
fo = "fo"
Uest = "Uest"
Dows = "Dows"
aquire = "aquire"

# typos that are programmatically observable
compileable = "compileable"
overlayed = "overlayed"

[default]
extend-ignore-identifiers-re = ["^[a-zA-Z][a-zA-Z]?[a-zA-Z]?[a-zA-Z]?$"]
```

</details>

### CODE REVIEW

**Comment by DilumAluthge** [comment]
I've added `crate-ci/typos@*` to the GitHub Actions allowlist for this repo.

File: .github/workflows/SpellCheck.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+on: [pull_request]
+
+jobs:
+  typos-check:
+    name: Spell Check with Typos
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Actions Repository
+        uses: actions/checkout@v4
```
Comment (DilumAluthge) [inline .github/workflows/SpellCheck.yml:13]:
```suggestion
        uses: actions/checkout@v4
        with:
          persist-credentials: false
```

**Comment by LilithHafner** [comment]
I was wondering what was going on :)

I noticed you asked for a short timeout in https://github.com/JuliaLang/julia/pull/48359, do you want that here too?

**Comment by DilumAluthge** [comment]
> I noticed you asked for a short timeout in #48359, do you want that here too?

Yeah, that would be good. Looks like this action runs pretty quickly, so a few minutes should be sufficient for a timeout.

File: .github/workflows/SpellCheck.yml (hunk 1 above)
Comment (LilithHafner) [reply .github/workflows/SpellCheck.yml:13]: Jeepers; how does actions/checkout@v4 even get any credentials when there is a toplevel `permissions: {}`?

**Comment by DilumAluthge** [comment]
There's also codespell (https://github.com/codespell-project/codespell), which might have fewer false positives.

File: .github/workflows/SpellCheck.yml (hunk 1 above)
Comment (DilumAluthge) [reply .github/workflows/SpellCheck.yml:13]: IIRC, if you pass an empty dict, it populates it with some default values.

Comment (DilumAluthge) [reply .github/workflows/SpellCheck.yml:13]: I might be misremembering.

Comment (DilumAluthge) [reply .github/workflows/SpellCheck.yml:13]: Just to be safe, let's be explicit:

```yaml
permissions:
  contents: read
```

File: .github/workflows/SpellCheck.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+name: Spell Check
+
+permissions: {}
```
Comment (DilumAluthge) [inline .github/workflows/SpellCheck.yml:3]:
```suggestion
permissions:
  contents: read
```

File: .github/workflows/SpellCheck.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  contents: read
+
+on: [pull_request]
+
+jobs:
+  typos-check:
+    name: Spell Check with Typos
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout Actions Repository
```
Comment (DilumAluthge) [inline .github/workflows/SpellCheck.yml:13]:
```suggestion
      - name: Checkout the JuliaLang/julia repository
```

File: .github/workflows/SpellCheck.yml [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+permissions:
+  contents: read
+
+on: [pull_request]
+
+jobs:
+  typos-check:
+    name: Spell Check with Typos
+    runs-on: ubuntu-latest
```
Comment (DilumAluthge) [inline .github/workflows/SpellCheck.yml:11]:
```suggestion
    runs-on: ubuntu-latest
    timeout-minutes: 5
```

File: .github/workflows/SpellCheck.yml (hunk 2 above)
Comment (LilithHafner) [reply .github/workflows/SpellCheck.yml]: `permissions: {}` is strictly less permissive (rmissions-to-jobs#defining-access-for-the-github_token-scopes). Why add these additional permissions?

File: .github/workflows/SpellCheck.yml [hunk 5]
```diff
@@ -0,0 +1,19 @@
+name: Spell Check
+
+permissions:
+  contents: read
```
Comment (DilumAluthge) [inline .github/workflows/SpellCheck.yml:4]:
```suggestion
permissions: {}
```

File: .github/workflows/SpellCheck.yml (hunk 2 above)
Comment (DilumAluthge) [reply .github/workflows/SpellCheck.yml:3]: Yeah, let's change it back to `permissions: {}`.

I'm not 100% sure how `actions/checkout` is able to get the contents of the repo even when we set `permissions: {}`. I wonder if the `permissions` key applies only to uses of the GitHub API (and doesn't apply to access over Git SSH).

Comment (LilithHafner) [reply .github/workflows/SpellCheck.yml:3]: The repo is public, they could be downloading with https.jl for all I know, lol.

Comment (DilumAluthge) [reply .github/workflows/SpellCheck.yml:3]: 😂 

For the API, GitHub rate-limits`https://api.github.com/*`, and the rate-limits are much lower if you don't provide a GitHub token.

I'm guessing that GitHub also has some rate-limiting or throttling in place for `https://github.com/*`.

**Review by DilumAluthge — COMMENTED** [review]
Overall, this seems like a good plan.

I'll do another thorough line-by-line review later.

We likely want to wait to merge this until the actual typos (the true positives) have been fixed. Lilith has a PR open for that.

File: .github/workflows/SpellCheck.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  typos-check:
+    name: (add false positives to typos.toml)
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+    steps:
+      - name: Checkout the JuliaLang/julia repository
+        uses: actions/checkout@v4
+        with:
+          persist-credentials: false
+      - name: Check spelling wit typos
```
Comment (LilithHafner) [inline .github/workflows/SpellCheck.yml:17]: Oh the irony.

```suggestion
      - name: Check spelling with typos
```

File: typos.toml [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+eyt = "eyt"
+matc = "matc"
+nam = "nam"
+alls = "alls"
+noe = "noe"
+
+# strings in tests directories
+fo = "fo"
+Uest = "Uest"
+Dows = "Dows"
```
Comment (szepeviktor) [inline typos.toml:88]: These should be in `extend-ignore-identifiers-re` with some context.
```toml
extend-ignore-identifiers-re = [
    "^[a-zA-Z][a-zA-Z]?[a-zA-Z]?[a-zA-Z]?$",
    '"QU QUer QUest"',
    '"wiNDows"',
]
```

Comment (LilithHafner) [reply typos.toml:88]: I really just want to do [`extend-words-re = "^[a-zA-Z][a-zA-Z]?[a-zA-Z]?[a-zA-Z]?$"`](https://github.com/crate-ci/typos/issues/852) and delete all of the short entries in extend-words.

**Comment by LilithHafner** [comment]
The presence of false positives, even with a fairly easy way of adding false positives to the list of exclusions, makes me think this is too strict. I would rather have an action that computes the set of "typos" on a per-file basis both before and after the PR and makes a comment/fails a non-required test if the set expands during a PR. That way there is no action required in the event of a false positive.

**Comment by szepeviktor** [comment]
`crate-ci/typos` is not the best action in the world.
If you run `typos` binary yourself you can say
```bash
typos || true 'Annotate only'
```
to have GitHub check annotations but never fail the workflow.

**Comment by szepeviktor** [comment]
I encourage every Julia author to leave the old way behind, distance themselves from assembler mnemonic-like variable names.

**Comment by LilithHafner** [comment]
Most false positives are not bad variable names. Here's a selection of false positives:

```
error: `egal` should be `equal`
  --> ./src/jlapi.c:437:22
    |
437 | JL_DLLEXPORT int ijl_egal(jl_value_t *a, jl_value_t *b)
    |                      ^^^^
    |
error: `egal` should be `equal`
  --> ./src/jlapi.c:439:15
    |
439 |     return jl_egal(a, b);
    |               ^^^^
    |
error: `Alls` should be `All`, `Falls`
  --> ./src/subtype.c:81:64
   |
81 |     // array of typevars that our bounds depend on, whose UnionAlls need to be
   |                                                                ^^^^
   |
error: `fo` should be `of`, `for`, `do`, `go`, `to`
  --> ./test/strings/search.jl:346:18
    |
346 | @test findnext(r"fo", "foo,bar,baz", 3) == nothing
    |                  ^^
    |
error: `ba` should be `by`, `be`
  --> ./test/strings/basic.jl:508:85
    |
508 |     @test_throws ArgumentError Base.unsafe_convert(Cstring, Base.cconvert(Cstring, "ba\0d"))
    |                                                                                     ^^
    |
error: `bage` should be `bag`
  --> ./test/channels.jl:347:61
    |
347 |     @noinline garbage_finalizer(f) = (finalizer(f, "gar" * "bage"); nothing)
    |                                                             ^^^^
    |
error: `ba` should be `by`, `be`
  --> ./test/broadcast.jl:586:15
    |
586 |         @test ba isa BitArray
    |               ^^
    |
```

To see all 4,000 false positives when using the default `typos` dictionary, clone the repo, checkout this PR, delete typos.toml, and run `typos`

**Comment by szepeviktor** [comment]
Could you help me? What is the meaning of `jl_egal`?

`julia_equival`?

**Comment by szepeviktor** [comment]
```
git diff-index --name-only --diff-filter=d -z master
```

Here `HEAD` may be the correct ref, not `master`.

File: .github/workflows/SpellCheck.yml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+          RELEASE_ASSET_URL="$(
+            gh api /repos/crate-ci/typos/releases/latest \
+            --jq '."assets"[] | select(."name" | test("^typos-.+-x86_64-unknown-linux-musl\\.tar\\.gz$")) | ."browser_download_url"'
+          )"
+          wget --secure-protocol=TLSv1_3 --max-redirect=1 --retry-on-host-error --retry-connrefused --tries=3 \
+            --quiet --output-document=- "${RELEASE_ASSET_URL}" \
+            | tar -xz -C "${{ runner.temp }}/typos" ./typos
+          "${{ runner.temp }}/typos/typos" --version
+          git diff-index --name-only --diff-filter=d -z ${{ github.base_ref }} \
+            | xargs -0"${{ runner.temp }}/typos/typos" --format json \
```
Comment (szepeviktor) [inline .github/workflows/SpellCheck.yml:33]:
```suggestion
            | xargs -0 -- "${{ runner.temp }}/typos/typos" --format json \
```

**Comment by szepeviktor** [comment]
Here you go ✈️

image

**Comment by LilithHafner** [comment]
> Could you help me? What is the meaning of `jl_egal`?

egal (`===`) is stronger than equal (`==`)

https://docs.julialang.org/en/v1/base/base/#Core.:===

> `julia_equal`?

In the context of the Julia compiler, I think it is a good idea to abbreviate `julia` as `jl`. Concision has some value, and it's entirely unambiguous what `jl` stands for.

File: .github/workflows/SpellCheck.yml (hunk 8 above)
Comment (LilithHafner) [reply .github/workflows/SpellCheck.yml]: `--` is unnecessary.

**Comment by szepeviktor** [comment]
> egal (`===`) is stronger than equal (`==`)

Thank you!

Is "egal" in German?
https://www.duden.de/rechtschreibung/egal_gleich_einerlei

**Comment by epage** [comment]
btw I'm hoping a good way to handle old vs new typos is Sarif integration (crate-ci/typos#594).

This is a format that allows "linters" to report issues to github and it handles tracking a backlog of them.  I don't know too much of the backlog side of this as I use this for `cargo clippy` and resolve out all of my SARIF lints before merging.

**Comment by LilithHafner** [comment]
In addition to the reasons listed in https://github.com/JuliaLang/julia/pull/51706#issuecomment-1763143908, I prefer `typos` over `codespell` because of the `typos` developers' quick responsiveness. (e.g. https://github.com/crate-ci/typos/pull/854)

**Comment by LilithHafner** [comment]
Some false positivity stats

Ignore words of length ≤ n | Typos before #51709 (X) | Typos after #51709 | Typos after #51709 excluding "compileable", "overlayed", "inconsistentcy", and "supress" which were not changed (Y) | False positives (Y) | True positives (X - Y) | True positive rate (X - Y) / X
--- | --- | --- | --- | --- | --- | ---
0 | 1349 | 1155 | 1091 | ≤1091 | ≥258 | ≥19.1%
1 | 1349 | 1155 | 1091 | ≤1091 | ≥258 | ≥19.1%
2 | 1032 | 838 | 774 | ≤774 | ≥258 | ≥25.0%
3 | 881 | 690 | 626 | ≤626 | ≥255 | ≥28.9%
4 | 355 | 168 | 104 | 104 | 251 | 70.7%
5 | 277 | 96 | 32 | 32 | 245 | 88.4%
6 | 254 | 79 | 15 | 15 | 239 | 94.1%
7 | 141 | 71 | 13 | 13 | 128 | 90.8%
8 | 122 | 61 | 3 | 3 | 119 | 97.5%
9 | 87 | 41 | 2 | 2 | 85 | 97.7%
10 | 58 | 40 | 1 | 1 | 57 | 98.3%
11 | 9 | 2 | 0 | 0 | 9 | 100.0%
12 | 5 | 2 | 0 | 0 | 5 | 100.0%
13 | 4 | 2 | 0 | 0 | 4 | 100.0%
14 | 1 | 0 | 0 | 0 | 1 | 100.0%
15 | 0 | 0 | 0 | 0 | 0 | N/A

Screenshot 2023-10-16 at 6 08 56 PM

The table is the right-handed cumulative sum while the plot represents the data for each length individually.

<details>
<summary>Some code I used</summary>

```julia

head = """[default]
extend-ignore-words-re = ["^"""
repeated = """[a-zA-Z]?"""
tail1 = """\$"]

[default.extend-words]
compileable = "compileable"
overlayed = "overlayed"
inconsistentcy = "inconsistentcy"
supress = "supress"
"""
tail2 = """\$"]
"""
function f(n, tail)
     open("typos.toml", "w") do io
          print(io, head)
          for i in 1:n
               print(io, repeated)
          end
          print(io, tail)
     end
     buf = IOBuffer()
     wait(run(pipeline(`../julia/typos --format brief`, `wc -l`, buf), wait=false))
     return parse(Int, String(take!(buf)))
end

```

</details>

The inequalities are because I may have missed some true typos of short words, but the vast majority seemed to be false positives so I didn't keep checking.

Informed by this data, I'm adding a regex that ignores typos with 5 or fewer characters.

**Comment by epage** [comment]
If any of the false positives from abbreviations are from the standard library, I'd be interested in getting those baked into a julia-specific dictionary. Similar for highly common package names (e.g. `flate2` in rust) or community api names (e.g. `ser` in rust). That can lead to a better experience for all Julia users.

**Comment by LilithHafner** [comment]
Here's a list of false positives that seem worth including in a Julia-specific dictionary:
```
Alls # All is a keyword
alls 
Egal # egal is a term for `===`
egal 
Egals
egals
Missings # Missing is a type (noun that can be pluralized)
missings # missing is a value (noun that can be pluralized)
ND # nd stands for number of dimensions, which is somewhat common
Nd
nd
somes # Some is a type (noun that can be pluralized)
splitted # `split` is a Base function so splitted is its past tense
modul # module and using are keywords, these abbreviations are used in Base and may be used elsewhere. The key detail is nobody will accidentally write modul or usig when they mean module or using because that would be caught much earlier than a typo check.
usig
whos # A function from before 1.0 that has since been removed but folks may still refer to it?
```
```
shttp # is an actuall protocall, and markdown blocks can be fenced with it as a label. This is not Julia specific.
```
Almost all of the remaining are abbreviations that happen to look like typos or deliberate typos used in testing.

**Comment by epage** [comment]
> Alls # All is a keyword
> alls 
> Missings # Missing is a type (noun that can be pluralized)
> missings # missing is a value (noun that can be pluralized)
> somes # Some is a type (noun that can be pluralized)
> splitted # `split` is a Base function so splitted is its past tense

Rust also has `Some` but I never see `somes` used.  Is that commonly in code or is it more a thing in documentation, like
```rust
// Here are several Somes
```
In Rust, I'd expect to see that written as
```rust
// Here are several `Some`s
```
which negates the problem.

> ND # nd stands for number of dimensions, which is somewhat common
> Nd
> nd

For Python, we specifically approved `NDArray`.  I was looking in docs, trying to find how this is used in Julia but couldn't.  Mind giving me some pointers?

> modul # module and using are keywords, these abbreviations are used in Base and may be used elsewhere. The key detail is nobody will accidentally write modul or usig when they mean module or using because that would be caught much earlier than a typo check.
> usig

I'm assuming this is the same problem as Rust when we need to refer to a `crate` (keyword), we fallback to `krate` and these are the fully idomatic ways of doing so.

**Comment by LilithHafner** [comment]
`somes` comes up in exactly one code snippet:
```julia
# Test completion for a case when type inference returned `Union` of the same types
union_somes(a, b) = rand() < 0.5 ? Some(a) : Some(b)
let s = "union_somes(1, 1.0)."
    c, r, res = test_complete_context(s)
    @test res
    @test "value" in c
end
```
```julia
function _reverse(A::BitArray, d::Int)
    nd = ndims(A)
    1 ≤ d ≤ nd || throw(ArgumentError("dimension $d is not 1 ≤ $d ≤ $nd"))
```

**Comment by LilithHafner** [comment]
The design now (which I think may be the final design) is...

Look at each file that was edited and neither created nor deleted. Scan these all for typos and compute the union. This is our baseline of "false positives". Compile this list into a Python set.

Look at each file that was edited and not delegated. Scan these all for typos, skipping any typos that are in the set of false positives. For each remaining typo, provide a github annotation and, if the typo is longer than 5 characters (at least 94% true positivity rate for typos that long), mark the CI job as a failure but continue to report all remaining novel typos.

This achieves:
- Close to the full sensitivity of the `typos` crate
- Almost never erroneously fails CI
- Few false positives reported (hopefully)
- When there is a false positive, it is safe to simply ignore it and press merge, even if the typos CI check was failing.

Runtime: 6s total, 1s doing the actual checks, `typos` goes through only the files that have been changed (and goes through them twice) bash handles lists of length equal to the number of files edited, python runs in O(number of reported typos). 5 minute timeout.

**Comment by KristofferC** [comment]
Looking at the suggestions in https://github.com/JuliaLang/julia/pull/51707, spell-checking actual running code seems like a bad idea. Running it on comments and other types of text seems useful.

**Comment by LilithHafner** [comment]
If it were trivial to exclude running code, I would have done so. However, I do still think it is acceptable to spell check running code. Specifically, "overlayed" "compileable" and "supress" have all made it into running code. `supress_output` actually made it into public API. Further, false positive avoidance seems pretty good to me. None of the suggestions you were laughing at in #51707 would be flagged in this PR.

<details>
<summary>Usage of "overlayed" "compileable" and "supress"</summary>

```
test/namedtuple.jl
390:            @test eff.nonoverlayed
397:            @test eff.nonoverlayed

test/reflection.jl
1030:        @test !Core.Compiler.is_nonoverlayed(effects)

test/compiler/codegen.jl
799:@test Core.Compiler.get_compileable_sig(which(f48085, (Vararg{Any},)), Tuple{typeof(f48085), Vararg{Int}}, Core.svec()) === nothing
800:@test Core.Compiler.get_compileable_sig(which(f48085, (Vararg{Any},)), Tuple{typeof(f48085), Int, Vararg{Int}}, Core.svec()) === Tuple{typeof(f48085), Any, Vararg{Any}}

test/compiler/AbstractInterpreter.jl
23:@MethodTable OverlayedMT
24:CC.method_table(interp::MTOverlayInterp) = CC.OverlayMethodTable(CC.get_world_counter(interp), OverlayedMT)
35:@overlay OverlayedMT strangesin(x::Float64) = iszero(x) ? nothing : cos(x)
37:# inference should use the overlayed method table
45:# effect analysis should figure out that the overlayed method is used
48:end |> !Core.Compiler.is_nonoverlayed
51:end |> !Core.Compiler.is_nonoverlayed
62:        @test Base.infer_effects(callstrange_entry, (Any,); interp) |> !Core.Compiler.is_nonoverlayed
65:        @warn "`nonoverlayed` test for inference bailing out is skipped since the method match sort order is changed."
72:end |> Core.Compiler.is_nonoverlayed
75:end |> Core.Compiler.is_nonoverlayed
87:@overlay OverlayedMT overlay_match(::Int) = missing
109:# disable partial concrete evaluation when tainted by any overlayed call
141:# Should not concrete-eval overlayed methods in semi-concrete interpretation

stdlib/REPL/src/TerminalMenus/config.jl
130: - `supress_output::Bool=false`: Ignored legacy argument, pass `suppress_output` as a keyword argument to `request` instead.
144:                supress_output::Union{Nothing, Bool}=nothing,   # typo was documented, unfortunately
176:    supress_output isa Bool   && (CONFIG[:supress_output] = supress_output)
183:config(charset=:ascii, scroll=:nowrap, supress_output=false, ctrl_c_interrupt=true)

base/loading.jl
2306:        exit(125) # we define status = 125 means PrecompileableError

base/compiler/utilities.jl
157:function get_compileable_sig(method::Method, @nospecialize(atype), sparams::SimpleVector)
162:        mt, atype, sparams, method, #=int return_if_compileable=#1)
170:        mt, atype, sparams, method, #=int return_if_compileable=#0)
173:isa_compileable_sig(@nospecialize(atype), sparams::SimpleVector, method::Method) =
174:    !iszero(ccall(:jl_isa_compileable_sig, Int32, (Any, Any, Any), atype, sparams, method))

base/compiler/typeinfer.jl
349:        cache_the_tree = ci.inferred && (is_inlineable(ci) || isa_compileable_sig(linfo.specTypes, linfo.sparam_vals, def))

base/compiler/abstractinterpretation.jl
165:            csig = get_compileable_sig(method, sig, match.sparams)
454:        # ignore the `:nonoverlayed` property if `interp` doesn't use overlayed method table
456:        if !isoverlayed(method_table(interp))
457:            all_effects = Effects(all_effects; nonoverlayed=false)
848:            if is_nonoverlayed(interp) || is_nonoverlayed(effects)
851:            # disable concrete-evaluation if this function call is tainted by some overlayed
852:            # method since currently there is no easy way to execute overlayed methods
853:            add_remark!(interp, sv, "[constprop] Concrete eval disabled for overlayed methods")

base/compiler/ssair/irinterp.jl
18:        (is_nonoverlayed(interp) || is_nonoverlayed(effects)))

base/compiler/ssair/inlining.jl
812:function compileable_specialization(mi::MethodInstance, effects::Effects,
817:        new_atype = get_compileable_sig(method, atype, sparams)
839:function compileable_specialization(match::MethodMatch, effects::Effects,
842:    return compileable_specialization(mi, effects, et, info; compilesig_invokes)
894:        return compileable_specialization(mi, effects, et, info;
899:    src === nothing && return compileable_specialization(mi, effects, et, info;
1501:        return compileable_specialization(mi, result.effects, et, info;
1535:        return compileable_specialization(result.mi, result.effects, et, info;
1594:    case = compileable_specialization(match, Effects(), InliningEdgeTracker(state), info;

base/compiler/methodtable.jl
66:`overlayed` indicates if any of the matching methods comes from an overlayed method table.
115:        (match::Union{MethodMatch,Nothing}, valid_worlds::WorldRange, overlayed::Bool)
129:`overlayed` indicates if any of the matching methods comes from an overlayed method table.

base/compiler/effects.jl
45:- `nonoverlayed::Bool`: indicates that any methods that may be called within this method
46:  are not defined in an [overlayed method table](@ref OverlayMethodTable).
94:Additionally, if the `nonoverlayed` property is false, a red prime symbol (′) is displayed after the tuple.
104:    nonoverlayed::Bool
113:        nonoverlayed::Bool)
122:            nonoverlayed)
151:const EFFECTS_UNKNOWN  = Effects(ALWAYS_FALSE, ALWAYS_FALSE, false, false, false, ALWAYS_FALSE, ALWAYS_FALSE, true) # unknown mostly, but it's not overlayed at least (e.g. it's not a call)
162:    nonoverlayed::Bool = effects.nonoverlayed)
171:        nonoverlayed)
183:        merge_effectbits(old.nonoverlayed, new.nonoverlayed))
200:is_nonoverlayed(effects::Effects)        = effects.nonoverlayed
241:           ((e.nonoverlayed        % UInt32) << 12)

base/compiler/ssair/show.jl
1021:    e.nonoverlayed || printstyled(io, '′'; color=:red)

base/compiler/inferencestate.jl
302:            ipo_effects = Effects(ipo_effects; nonoverlayed=is_nonoverlayed(def))
323:is_nonoverlayed(m::Method) = !isdefined(m, :external_mt)
324:is_nonoverlayed(interp::AbstractInterpreter) = !isoverlayed(method_table(interp))
325:isoverlayed(::MethodTableView) = error("unsatisfied MethodTableView interface")
326:isoverlayed(::InternalMethodTable) = false
327:isoverlayed(::OverlayMethodTable) = true
328:isoverlayed(mt::CachedMethodTable) = isoverlayed(mt.table)

src/gf.c
802:            // there are insufficient given parameters for jl_isa_compileable_sig now to like this type
1004:JL_DLLEXPORT int jl_isa_compileable_sig(
1279:        // In most cases `!jl_isa_compileable_sig(tt, sparams, definition)`,
1291:    // TODO: maybe assert(jl_isa_compileable_sig(compilationsig, sparams, definition));
2591:                                                        int return_if_compileable)
2599:    int is_compileable = ((jl_datatype_t*)ti)->isdispatchtuple;
2602:        if (!is_compileable) {
2612:    if (!is_compileable)
2613:        is_compileable = jl_isa_compileable_sig(tt, env, m);
2615:    return (!return_if_compileable || is_compileable) ? (jl_value_t*)tt : jl_nothing;
2638:// return a MethodInstance for a compileable method_match
2702:// tries to find a method for which the requested signature is compileable.
2727:        // first, select methods for which `types` is compileable
2731:            if (jl_isa_compileable_sig(types, match1->sparams, match1->method))

src/julia.h
435:    //     uint8_t ipo_nonoverlayed        : 1;
444:    //     uint8_t nonoverlayed        : 1;
1564:JL_DLLEXPORT int jl_isa_compileable_sig(jl_tupletype_t *type, jl_svec_t *sparams, jl_method_t *definition);

src/precompile_utils.c
160:        if (jl_is_datatype(m->sig) && jl_isa_compileable_sig((jl_tupletype_t*)m->sig, jl_emptysvec, m)) {
252:            if (mi != jl_atomic_load_relaxed(&mi->def.method->unspecialized) && !jl_isa_compileable_sig((jl_tupletype_t*)mi->specTypes, mi->sparam_vals, mi->def.method))

src/jl_exported_funcs.inc
288:    XX(jl_isa_compileable_sig) \
```

</details>

It's hard to know how much false positives will be an issue without trying and finding out, but I suspect that the false positivity rate will be very low in practice. It can be bounded above by the reported false positivity rates in https://github.com/JuliaLang/julia/pull/51704#issuecomment-1765444105, assuming PR text is more typo-ey than text already merged into master.

**Comment by LilithHafner** [comment]
See #51739 for an example of this action applied to the most recent PR with significant code additions.

Feel free to recommend any other open PR for me to test this out on. I'm optimistic it will perform well.

**Comment by LilithHafner** [comment]
I'm going to remove the test typos and get this ready for merge. We can always revert if there are too many false positives.

**Comment by KristofferC** [comment]
> However, I do still think it is acceptable to spell check running code. Specifically, "overlayed" "compileable" and "supress" have all made it into running code. supress_output actually made it into public API.

That are indeed good reasons why running it on code is useful.

**Comment by LilithHafner** [comment]
Here's another example of catching a code typo

https://github.com/JuliaLang/julia/pull/52006/commits/1d904873598e9ecf7fd25f937adabae43b59c05f
